File: Solutions/Minimum_Number_of_Arrows_to_Burst_Balloons/Minimum_Number_of_Arrows_to_Burst_Balloons.go
//Most common
func findMinArrowShots(points [][]int) int {
	sort.Slice(points, func(i, j int) bool {
		return points[i][1] < points[j][1]
	})
	shoot := 0
	count := 0
	i := 0
	for i < len(points) {
		shoot = points[i][1]
		count++
		for i < len(points) && points[i][0] <= shoot && points[i][1] >= shoot {
			i++
		}
	}
	return count
}

//Second fastest
func findMinArrowShots(points [][]int) int {
    if len(points) < 2 {
        return len(points)
    }
    sort.Slice(points, func(i, j int) bool {
        return points[i][1] < points[j][1]
    })
    
    end := points[0][1]
    ans := 1
    for i := 1; i < len(points); i++ {
        if points[i][0] > end {
            end = points[i][1]
            ans++
        }
    }
    return ans
}


//Fastest
func findMinArrowShots(points [][]int) int {
	if len(points) == 0 {
		return 0
	}
	sort.Sort(shots(points))
	i := 0
	j := 1
	count := 1
	for {
		if i+j > len(points)-1 {
			break
		}
		end := points[i][1]
		if end >= points[i+j][0] {
			j++
		} else {
			count++
			i = i + j
			j = 1
		}
	}
	return count    
}

type shots [][]int

func (s shots) Len() int {
	return len(s)
}
func (s shots) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
	return
}
func (s shots) Less(i, j int) bool {
	if s[i][1] < s[j][1] {
		return true
	}
	return false
}

File: Solutions/Minimum_Number_of_Arrows_to_Burst_Balloons/Minimum_Number_of_Arrows_to_Burst_Balloons.py
class Solution:
    def findMinArrowShots(self, points):
        points.sort(key = lambda x: x[1])
        n, count = len(points), 1
        if n == 0: return 0
        curr = points[0]
        
        for i in range(n):
            if curr[1] < points[i][0]:
                count += 1
                curr = points[i]
                
        return count  

#Third fastest
class Solution:
    # sort by the end, consider the first one a0, if others after it has a start less than a0[0], should be ok with the 
    # same arrow
    def findMinArrowShots(self, points: List[List[int]]) -> int:
        
        points.sort(key=lambda x : x[1])
        n = 0
        cur = -sys.maxsize - 1
        for p in points:
            if p[0] <= cur:
                continue
            n += 1
            cur = p[1]
        return n

#Second fastest
class Solution:
  def findMinArrowShots(self, points: List[List[int]]) -> int:
    if not points:
      return 0
    
    points.sort(key=lambda x:x[1])
    _, right_bound = points[0]
    # print(points)
    ans = 0
    for point in points[1:]:
      if point[0] > right_bound:
        _, right_bound = point
        ans += 1
    return ans + 1

#Fastest
class Solution:
    def findMinArrowShots(self, points: List[List[int]]) -> int:
        if not points:
            return 0
        points.sort(key=lambda x:x[1])
        count=1
        begin=points[0][1]
        for start,end in points:
            if start >begin:
                count+=1
                begin=end
        return count
